### PR TITLE
made randomised computer players

### PR DESCRIPTION
**Pull request by iatkins**

### CODE REVIEW

File: server/controllers/rooms.js [hunk 1]
```diff
@@ -56,6 +56,24 @@ router.post('/join', auth, (req, res) => {
     }
 });
 
+/* Add an automated player to the room */
+router.post('/addAutoPlayer', (req, res) => {
+    try {
+        const { roomId }= req.body;
```
Comment (tillywoodfield) [inline server/controllers/rooms.js]: `const { roomId } = req.body;`

File: server/controllers/rooms.js [hunk 2]
```diff
@@ -56,6 +56,24 @@ router.post('/join', auth, (req, res) => {
...
+        const x = Room.getById(roomId).gameState.getNumberofAutoPlayers();
```
Comment (tillywoodfield) [inline server/controllers/rooms.js]: Rename `x` to be more descriptive? e.g. numAutoPlayers? Also can `getNumberofAutoPlayers();` have a capital O for Of

File: server/controllers/rooms.js [hunk 3]
```diff
@@ -56,6 +56,24 @@ router.post('/join', auth, (req, res) => {
...
+        const user = { username: "Computer " + x, id: 0, real: false};
+        const room = Room.getById(roomId);
```
Comment (tillywoodfield) [inline server/controllers/rooms.js:65]: I'm not sure this variable is being used?

Comment (tillywoodfield) [reply server/controllers/rooms.js:65]: Oh never mind

File: server/models/GameLogic.js [hunk 4]
```diff
@@ -2,6 +2,7 @@ const cardsManager = require('../services/cardsManager');
 const validWord = require('../services/validWord');
 const socket = require('../services/socket');
 const { statusTypes } = require('./statusTypes');
+const AI = require('../services/AI');
```
Comment (tillywoodfield) [inline server/models/GameLogic.js:4]: Not sure how I feel about capital file names? @Ro4052 ?

File: server/models/GameLogic.js [hunk 5]
```diff
@@ -65,6 +66,8 @@ class GameLogic {
     /* Return the list of players, hiding their assigned cards */
     getPlayers() { return this.players.map(player => ({ username: player.username, id: player.id, score: player.score, finishedTurn: player.finishedTurn })) };
 
+    getNumberofAutoPlayers() { return this.players.filter(player => player.real === false).length };
```
Comment (tillywoodfield) [inline server/models/GameLogic.js]: Change `player.real === false` to `!player.real`

File: server/models/GameLogic.js [hunk 6]
```diff
@@ -233,7 +249,7 @@ class GameLogic {
     /* Adds player's card to list of played cards */
     playCard(username, cardId) {
         if (this.status !== statusTypes.WAITING_FOR_OTHER_PLAYERS) {
-            throw Error("Now is not the right time to play a card");
+            throw Error("Now is not the right time to play a card" + username);
```
Comment (tillywoodfield) [inline server/models/GameLogic.js]: Remove the added username

File: server/models/GameLogic.js [hunk 7]
```diff
@@ -262,6 +278,13 @@ class GameLogic {
         this.shufflePlayedCards();
         socket.emitPlayedCards(this.roomId, this.getPlayedCards());
         socket.promptPlayersVote(this.roomId, this.currentPlayer, voteDuration);
+        this.players.forEach(player => {
```
Comment (tillywoodfield) [inline server/models/GameLogic.js:298]: Can this for loop be moved into it's own function like like play card for ais or something?

Comment (tillywoodfield) [reply server/models/GameLogic.js:298]: And anywhere else that has a similar loop

File: server/services/AI.js [hunk 8]
```diff
@@ -0,0 +1,11 @@
+// where the maths will be, ie functionality for the AI person
+
+/* Random card picked*/
+exports.autoPickCard = cards => {
+    const randomCard = cards[Math.floor(Math.random()*cards.length)];
+    return cardId = randomCard.cardId;
```
Comment (tillywoodfield) [inline server/services/AI.js]: `return randomCard.cardId;`

File: server/services/AI.js [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+}
+
+exports.autoWord = () => {
+    return word = "hello";
```
Comment (tillywoodfield) [inline server/services/AI.js]: `return "hello";`

File: src/components/Lobby/LobbyActions.js [hunk 10]
```diff
@@ -33,3 +38,11 @@ export const leaveRoom = roomId => dispatch => {
     })
     .catch(err => dispatch(joinRoomFailure(err.message)));
 };
+
+export const addAutoPlayer = roomId => dispatch => {
```
Comment (tillywoodfield) [inline src/components/Lobby/LobbyActions.js]: Possibly might be better as calling them AI players than auto players?

**Comment by tillywoodfield** [comment]
Can you add e2e tests for all the auto stuff please?

File: server/models/GameLogic.js [hunk 11]
```diff
@@ -144,6 +147,9 @@ class GameLogic {
             this.currentPlayer = this.players[this.roundNum % this.players.length];
             socket.emitNewRound(this.roomId, this.status, this.roundNum, this.currentPlayer, storytellerDuration);
             this.nextRoundTimeout = setTimeout(this.nextRound.bind(this), storytellerDuration);
+            if (!this.currentPlayer.real) {
+            this.AIsPlayCardAndWord();
```
Comment (tillywoodfield) [inline server/models/GameLogic.js]: indent

Comment (tillywoodfield) [reply server/models/GameLogic.js]: It looks like you're trying to play before it gets to the next round?

**Review by tillywoodfield — DISMISSED** [review]

**Review by sibela46 — APPROVED** [review]

**Pull request merged**